server/ec2spotmanager/CloudProvider/GCECloudProvider.py

```diff
@@ -0,0 +1,376 @@
+'''
+GCE Cloud Provider
+
+@author:     Jesse Schwartzentruber (:truber)
+
+@license:
+
+This Source Code Form is subject to the terms of the Mozilla Public
+License, v. 2.0. If a copy of the MPL was not distributed with this
+file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+@contact:    truber@mozilla.com
+'''
+import logging
+import requests
+import yaml
+from django.conf import settings
+from laniakea.core.providers.gce import ComputeEngineManager
+from .CloudProvider import (CloudProvider, INSTANCE_STATE, INSTANCE_STATE_CODE, wrap_provider_errors)
+from ..tasks import SPOTMGR_TAG
+from ..common.gce import CORES_PER_INSTANCE, RAM_PER_INSTANCE
+
+
+#
+# Google Compute Platform provider for EC2SpotManager
+#
+# Currently laniakea is using libcloud as the backend, which means operations are synchronous.
+# API access is not very optimal as a result. Launching typically takes ~30s and terminating ~80s.
+#
+
+
+class _LowercaseDict(dict):
+
+    def __init__(self, *args, **kwds):
+        super(_LowercaseDict, self).__init__()
+        if len(args) > 1:
+            raise TypeError("dict expected at most 1 arguments, got %d" % (len(args),))
+        if args:
+            if hasattr(args[0], "items"):
+                for (k, v) in args[0].items():
+                    self[k] = v
+            else:
+                for (k, v) in args[0]:
+                    self[k] = v
+        elif kwds:
+            for (k, v) in kwds.items():
+                self[k] = v
+
+    def __getitem__(self, key):
+        return super(_LowercaseDict, self).__getitem__(key.lower())
+
+    def __setitem__(self, key, value):
+        return super(_LowercaseDict, self).__setitem__(key.lower(), value.lower())
+
+    def __delitem__(self, key):
+        return super(_LowercaseDict, self).__delitem__(key.lower())
+
+
+class GCECloudProvider(CloudProvider):
+    '''Implementation of CloudProvider interface for GCE.
+    '''
+    NODE_STATE_MAP = {
+        "PROVISIONING": INSTANCE_STATE["pending"],
```

Comment:
We could map this to INSTANCE_STATE["requested"] as well, right? Or does PROVISIONING mean that the request is already fulfilled?

Reply:
Probably yes. That's a translation from libcloud here: https://github.com/apache/libcloud/blob/trunk/libcloud/compute/drivers/gce.py#L1728

The point is moot though, since we only see Nodes returned once they are in the running state. I opened https://github.com/MozillaSecurity/laniakea/issues/41 about that.

Reply:
To be clear though, I will change that to "requested" ... I think you're right, it's the better representation in our model.